saliency/metrics/pic.py

```diff
@@ -0,0 +1,377 @@
+  blurred_images = []
+  predictions = []
+
+  # This list will contain mapping of image entropy for a given saliency
+  # threshold to model prediction.
+  entropy_pred_tuples = []
+
+  # Estimate entropy of the original image.
+  original_img_entropy = estimate_image_entropy(img)
+
+  # Estimate entropy of the completely blurred image.
+  fully_blurred_img = create_blurred_image(full_img=img, pixel_mask=random_mask)
+  fully_blurred_img_entropy = estimate_image_entropy(fully_blurred_img)
+
+  # Compute model prediction for the original image.
+  original_img_pred = pred_func(img[np.newaxis, ...])[0]
+
+  # Compute model prediction for the completely blurred image.
+  fully_blurred_img_pred = pred_func(fully_blurred_img[np.newaxis, ...])[0]
+
+  blurred_images.append(fully_blurred_img)
+  predictions.append(fully_blurred_img_pred)
+
+  # If the entropy of the completely blurred image is higher or equal to the
+  # entropy of the original image then the metric cannot be used for this
+  # image. Don't include this image in the aggregated result.
+  if fully_blurred_img_entropy >= original_img_entropy:
+    return None
+
+  # If the score of the model on completely blurred image is higher or equal to
+  # the score of the model on the original image then the metric cannot be used
+  # for this image. Don't include this image in the aggregated result.
+  if fully_blurred_img_pred >= original_img_pred:
+    return None
+
+  # Iterate through saliency thresholds and compute prediction of the model
+  # for the corresponding blurred images with the saliency pixels revealed.
+  max_normalized_pred = 0.0
+  for threshold in saliency_thresholds:
+    quantile = np.quantile(saliency_map, 1 - threshold)
+    pixel_mask = saliency_map >= quantile
+    pixel_mask = np.logical_or(pixel_mask, random_mask)
+    blurred_image = create_blurred_image(full_img=img, pixel_mask=pixel_mask)
+    entropy = estimate_image_entropy(blurred_image)
+    pred = pred_func(blurred_image[np.newaxis, ...])[0]
+    # Normalize the values, so they lie in [0, 1] interval.
+    normalized_entropy = (entropy - fully_blurred_img_entropy) / (
+        original_img_entropy - fully_blurred_img_entropy)
+    normalized_entropy = np.clip(normalized_entropy, 0.0, 1.0)
+    normalized_pred = (pred - fully_blurred_img_pred) / (
+        original_img_pred - fully_blurred_img_pred)
+    normalized_pred = np.clip(normalized_pred, 0.0, 1.0)
+    max_normalized_pred = max(max_normalized_pred, normalized_pred)
+
+    # Make normalized_pred only grow if keep_monotonous is true.
+    if keep_monotonous:
+      entropy_pred_tuples.append((normalized_entropy, max_normalized_pred))
+    else:
+      entropy_pred_tuples.append((normalized_entropy, normalized_pred))
+
+    blurred_images.append(blurred_image)
+    predictions.append(pred)
+
+  # Interpolate the PIC curve.
+  entropy_pred_tuples.append((0.0, 0.0))
+  entropy_pred_tuples.append((1.0, 1.0))
+
+  data_points = zip(*entropy_pred_tuples)
```

Comment:
a potentially clearer way to write this:


(entropy_data, pred_data) = zip(*entropy_pred_tuples)
interp_func = interpolate.interp1d(x=entropy_data, y=pred_data)